src/shims/unix/fd.rs

```diff
@@ -227,20 +223,25 @@ impl VisitProvenance for FdTable {
 }
 
 impl FdTable {
-    pub(crate) fn new(mute_stdout_stderr: bool) -> FdTable {
-        let mut fds: BTreeMap<_, FileDescriptor> = BTreeMap::new();
-        fds.insert(0i32, FileDescriptor::new(io::stdin()));
+    fn new() -> Self {
+        FdTable { fds: BTreeMap::new() }
+    }
+    pub(crate) fn init(mute_stdout_stderr: bool) -> FdTable {
+        let mut fds = FdTable::new();
+        fds.insert_fd(io::stdin());
         if mute_stdout_stderr {
-            fds.insert(1i32, FileDescriptor::new(NullOutput));
-            fds.insert(2i32, FileDescriptor::new(NullOutput));
+            assert_eq!(fds.insert_fd(NullOutput), 1);
+            assert_eq!(fds.insert_fd(NullOutput), 2);
         } else {
-            fds.insert(1i32, FileDescriptor::new(io::stdout()));
-            fds.insert(2i32, FileDescriptor::new(io::stderr()));
+            assert_eq!(fds.insert_fd(io::stdout()), 1);
+            assert_eq!(fds.insert_fd(io::stderr()), 2);
         }
-        FdTable { fds }
+        fds
     }
 
-    pub fn insert_fd(&mut self, file_handle: FileDescriptor) -> i32 {
+    /// Insert a file descriptor to the FdTable.
+    pub fn insert_fd<T: FileDescription>(&mut self, fd: T) -> i32 {
```

Comment:
I'm not sure what the motivation is for this API change, but it seems odd to me. Note that `insert_fd_with_min_fd` wasn't changed, so now we have two equally-named APIs doing very different things. And `insert_fd_with_min_fd` can't be changed since indeed it is used to insert duplicates of existing file descriptions that shouldn't be wrapped in a new `Rc`.

> This change is needed by https://github.com/rust-lang/miri/pull/3712.

I don't understand how it is possible for this change to be *needed*. This change means that the API is now strictly less powerful than it was before. The only argument I can currently imagine for this change is ergonomics, and I don't agree that's worth it given the API inconsistencies this introduces.

IMO this should be reverted, both `insert_fd` and `insert_fd_with_min_fd` should take a `FileDescriptionRef`.

Reply:
The API didn't need to be as powerful as it was, so this seems good (restricting to the use case we need). It's also the only sane design I see for having each `FileDescriptionrRef` carry a unique id that we can use instead of its address.

Reply:
We considered do it in `FileDescriptionRef::new`, but then that needs access to the InterpCx or at least the `FdTable`.

Reply:
It is pure coincidence that we have no operation which creates a 2nd file descriptor for an existing file description, and calls this function. In fact arguably [this here](https://github.com/rust-lang/miri/blob/79285bb87fb959d72d538a9c5eeaf5aabcea1829/src/shims/unix/fd.rs#L307) *should* just call `insert_fd`.

And having two `insert_fd` functions that don't even take the same kind of argument seems like an obviously bad API to me, *way* worse than an API that is slightly more general than what we happen to need right now.

Creating a `FileDescriptionRef` from a `FileDescription` should be a completely independent operation from inserting a `FileDescriptionRef` into the FD table.

Reply:
> It's also the only sane design I see for having each FileDescriptionrRef carry a unique id that we can use instead of its address.

Just have `FileDescriptionRef::new` take a reference like `ecx: &mut MiriInterpCx<'tcx>` and then it can do the ID assignment.

Reply:
I guess we could call the functions `insert_new_fd` and `insert_existing_fd` and then we don't even have to change their implementation. (Though i think factoring `FileDescriptionRef::new` into a separate operation, even if it remains private, would still help to make the conceptual structure more clear.)

The PR description should have mentioned the point about `FileDescriptionRef::new` needing access to global state in https://github.com/rust-lang/miri/pull/3712, that would have avoided some confusion. :)